File: sdk/java/src/main/java/com/zorroa/zmlp/sdk/app/ProjectApp.java
package com.zorroa.zmlp.sdk.app;

import com.fasterxml.jackson.core.type.TypeReference;
import com.zorroa.zmlp.sdk.ZmlpClient;
import com.zorroa.zmlp.sdk.domain.PagedList;
import com.zorroa.zmlp.sdk.domain.project.Project;
import com.zorroa.zmlp.sdk.domain.project.ProjectFilter;
import com.zorroa.zmlp.sdk.domain.project.ProjectSpec;

import java.util.UUID;

public class ProjectApp {

    private ZmlpClient client;

    public ProjectApp(ZmlpClient client) {
        this.client = client;
    }

    /**
     * Create a new project.
     *
     * @param spec A ProjectSpec instance.
     * @return The created Project
     */
    public Project createProject(ProjectSpec spec) {
        return client.post("/api/v1/projects", spec, Project.class);
    }

    /**
     * Find a Project with its unique Id.
     *
     * @param id
     * @return The found Project
     */
    public Project getProject(UUID id) {
        return client.get("/api/v1/projects/" + id.toString(), null, Project.class);
    }

    /**
     * Find a Project with the given filter.  Filter must limit result count to 1.
     *
     * @param filter
     * @return The found Project
     */
    public Project findProject(ProjectFilter filter) {
        return client.post("/api/v1/projects/_findOne", filter, Project.class);
    }

    /**
     * @param filter
     * @return
     */
    public PagedList<Project> searchProjects(ProjectFilter filter) {
        return client.post("/api/v1/projects/_search", filter, new TypeReference<PagedList<Project>>() {
        });
    }
}

File: client/admin/pylib/zmlp_admin/app/project_app.py
from zmlp.entity import Project, ProjectTier

from ..entity.project import ProjectSize

__all__ = [
    'ProjectApp'
]


class ProjectApp(object):

    def __init__(self, app):
        self.app = app

    def get_project(self, pid):
        """
        Get a project by it's unique Id.

        Args:
            pid (str): The project Id.

        Returns:
            Project
        """
        return Project(self.app.client.get(f'/api/v1/projects/{pid}'))

    def create_project(self, name, tier=ProjectTier.ESSENTIALS, size=ProjectSize.SMALL, pid=None):
        """
        Create a project.

        Args:
            name (str): The name of the project.
            tier: (ProjectTier): The project tier.
            size (ProjectSize): Used to determine the default index settings.
            pid: (str): An optional project UUID, otherwise its randomly generated.

        Returns:
            Project: The newly created project.
        """
        body = {
            'name': name,
            'tier': tier.name,
            'size': size.name,
            'id': pid
        }
        return Project(self.app.client.post('/api/v1/projects', body))
